docs/design-docs/02-custom-root-reconciler-clusterrole.md

```diff
@@ -0,0 +1,133 @@
+# Custom `ClusterRole` assignments for `RootSync`s
+
+* Author(s): @tomasaschan
+* Approver: @sdowell, @janetkuo
+* Status: approved
+
+## Summary
+
+When creating a reconciler deployment for a `RootSync`, Config Sync currently [creates a
+`ClusterRoleBinding`] for the generated `ServiceAccount` granting it `cluster-admin`.
+
+This proposal introduces a mechanism to customize this behavior.
+
+[creates a `ClusterRoleBinding`]: https://github.com/GoogleContainerTools/kpt-config-sync/blob/199db6dbfa0b9eb1824925c8c687574de095a294/pkg/reconcilermanager/controllers/rootsync_controller.go#L789
+
+## Motivation
+
+If one wants to allow deploying resources from a source where one doesn't have full
+control of the contents of the upstream config source, and want to limit what another
+team or third party can accomplish in the cluster, one can imagine limiting what
+permissions are granted to the `ServiceAccount` running the reconciler.
+
+However, as Config Sync is currently granting `cluster-admin`, any custom role bindings
+are effectively ignored; the reconciler will, due to the binding added by Config Sync,
+have access to do _anything_ regardless.
+
+To follow the [principle of least privilege], one should ensure the reconciler only has
+access to deploy the expected resources.
+
+See also [#935].
+
+[principle of least privilege]: https://en.wikipedia.org/wiki/Principle_of_least_privilege
+[#935]: https://github.com/GoogleContainerTools/kpt-config-sync/issues/935
+
+## Design Overview
+
+By providing the name of a user-defined `ClusterRole` for a `RootSync`, a user can
+override which role Config Sync binds to. This configuration is exposed as a new field
+on `spec.overrides` for a `RootSync`:
+
+```yaml
+spec:
+  overrides:
+    clusterRole:
+      name: my-custom-role
+```
+
+This role is defaulted to `cluster-admin` in order to stay backwards compatible.
+
+For the case where a single `ClusterRole` is not expressive enough to configure the
+permissions a user want, you can instead set `disabled: true` on the override object:
+
+```yaml
+spec:
+  overrides:
+    clusterRole:
+      disabled: true
```

Comment:
Whoever has access to RootSync has access to grant themselves any permission they want by specifying the right cluster role. This has the problem of privilege escalation. However, given that RootSync has cluster-admin access before this change, this change isn't that dangerous, but could be surprising to admins when they grant limited access but also grant RootSync access to someone else. @tomasaschan what do you think about this caveat? Ideally the toggle should live outside of RootSync, but it's less of a problem for RootSync compared to RepoSync API.

Reply:
In the setup we run which motivates this feature this won't be an issue - we do not intend to grant access to create new syncs or update existing ones. In fact, the whole point is to keep all syncs in a repo we control, so that we can review the setup and ensure permissions are what they should be, while allowing them to sync config from repos we don't. In other words, this feature is what allows us to _prevent_ privilege escalation.

So I don't think there's a privilege escalation vector here that wasn't here already. If you can create or edit syncs, you can sync in config from an arbitrary source, and grant yourself whatever permissions you want.

As for RepoSyncs, in addition to all of the above, they will not be able to create cluster scoped resources, so any escalation will be even less limited.

Reply:
The privilege escalation problem is more on the RepoSync given that it has no permission to do anything by default. 

>  If you can create or edit syncs, you can sync in config from an arbitrary source, and grant yourself whatever permissions you want.

This isn't true for RepoSync. RepoSync editors don't have access to modify the RBAC rolebindings granted to the namespace reconciler SA.